src/vfs2.c

```diff
@@ -977,277 +1060,271 @@ static int compare_wal_headers(struct vfs2_wal_hdr a,
+	ihdr->basic[0].iChange += 1;
+	ihdr->basic[0].mxFrame = mx;
+	ihdr->basic[0].nPage = num_pages;
+	/* XXX byte order */
+	ihdr->basic[0].frame_cksums.cksum1 = ByteGetBe32(fhdr.cksum1);
+	ihdr->basic[0].frame_cksums.cksum2 = ByteGetBe32(fhdr.cksum2);
+	struct cksums sums = {};
+	update_cksums(native_magic(), (const void *)&ihdr->basic[0], 40, &sums);
+	ihdr->basic[0].cksums = sums;
+	ihdr->basic[1] = ihdr->basic[0];
+}
 
-	/* Collect memory allocations in one place to simplify the control
-	 * flow. A small amount of memory will be leaked if one of the later
-	 * allocations fails. */
-	char *fixed1 = sqlite3_malloc(data->orig->mxPathname + 1);
-	char *fixed2 = sqlite3_malloc(data->orig->mxPathname + 1);
-	sqlite3_file *phys1 = sqlite3_malloc(data->orig->szOsFile);
-	sqlite3_file *phys2 = sqlite3_malloc(data->orig->szOsFile);
-	if (fixed1 == NULL || fixed2 == NULL || phys1 == NULL ||
-	    phys2 == NULL) {
+static void restart_full_hdr(struct wal_index_full_hdr *ihdr, struct wal_hdr new_whdr)
+{
+	/* cf. walRestartHdr */
+	ihdr->basic[0].mxFrame = 0;
+	ihdr->basic[0].salts = new_whdr.salts;
+	struct cksums sums = {};
+	update_cksums(native_magic(), (const void *)&ihdr->basic[0], 40, &sums);
+	ihdr->basic[0].cksums = sums;
+	ihdr->basic[1] = ihdr->basic[0];
+	ihdr->nBackfill = 0;
+	ihdr->nBackfillAttempted = 0;
+}
+
+static uint32_t wal_cursor_from_size(uint32_t page_size, sqlite3_int64 size)
+{
+	sqlite3_int64 whdr_size = (sqlite3_int64)sizeof(struct wal_hdr);
+	if (size < whdr_size) {
+		return 0;
+	}
+	sqlite3_int64 x = (size - whdr_size) / ((sqlite3_int64)sizeof(struct wal_frame_hdr) + (sqlite3_int64)page_size);
+	return (uint32_t)x;
+}
+
+static sqlite3_int64 wal_offset_from_cursor(uint32_t page_size, uint32_t cursor)
+{
+	return (sqlite3_int64)sizeof(struct wal_hdr) + (sqlite3_int64)cursor * ((sqlite3_int64)sizeof(struct wal_frame_hdr) + (sqlite3_int64)page_size);
+}
+
+static int open_entry(struct common *common, const char *name, struct entry *e)
+{
+	sqlite3_vfs *v = common->orig;
+	int path_cap = v->mxPathname + 1;
+	int file_cap = v->szOsFile;
+	int rv;
+
+	*e = (struct entry){};
+	e->common = common;
+	sm_init(&e->wtx_sm, wtx_invariant, NULL, wtx_states, WTX_CLOSED);
+
+	e->main_db_name = sqlite3_malloc(path_cap);
+	if (e->main_db_name == NULL) {
 		return SQLITE_NOMEM;
 	}
-	xout->wal.wal_cur_fixed_name = fixed1;
-	xout->wal.wal_prev_fixed_name = fixed2;
-	memset(phys1, 0, (size_t)data->orig->szOsFile);
-	xout->orig = phys1;
-	memset(phys2, 0, (size_t)data->orig->szOsFile);
-	xout->wal.wal_prev = phys2;
-	xout->wal.moving_name = name;
-
-	/* Open the two physical WALs. */
-	strncpy(fixed1, name, (size_t)(dash - name));
-	fixed1[dash - name] = '\0';
-	strcat(fixed1, VFS2_WAL_FIXED_SUFFIX1);
-	int out_flags1;
-	rv = data->orig->xOpen(data->orig, fixed1, phys1, flags, &out_flags1);
+	strcpy(e->main_db_name, name);
+
+	e->wal_moving_name = sqlite3_malloc(path_cap);
+	if (e->wal_moving_name == NULL) {
+		return SQLITE_NOMEM;
```

Comment:
Here and in similar places in this function freeing up resources is missing.

Reply:
Ack, will fix

Reply:
I spoke too soon. The lack of a cleanup codepath in this function is intentional, I just forgot to document it. The reason is that we always set the pMethods field of the file vtable in vfs2_open, which means that SQLite will call vfs2_close on our file object even if opening it failed. We lean on that to clean up after open_entry when it failed partway through. I needed to fix the last few lines of set_up_entry to properly implement this though, see https://github.com/canonical/dqlite/pull/631/commits/4c8cb1dfc12ce0b3899c2b5599e44055a83537fe.

Reply:
ack.